llm/memory.go

```diff
@@ -437,3 +448,39 @@ func projectorMemoryRequirements(filename string) (weights, graphSize uint64) {
 
 	return weights, graphSize
 }
+
+// estimateKvCacheSize determines the memory required for K or V cache based on the quantization type
+func estimateKvCacheSize(cacheType string, numCtx, blockCount, embeddingHeadCount, headCountKV uint64, isEmbeddingModel bool) uint64 {
+	var bytesPerElement float64
+
+	if isEmbeddingModel && cacheType != "f16" && cacheType != "f32" {
+		cacheType = "f16" // Default to f16 for embedding models if an unsupported type is specified
+	}
+
+	switch cacheType {
+	case "f32", "fp32":
+		bytesPerElement = 4 // fp32
+	case "", "f16", "fp16":
+		bytesPerElement = 2 // fp16
+	case "q8_0":
+		bytesPerElement = 1 // 1/2 of fp16
+	case "q5_1":
+		bytesPerElement = 0.65
+	case "q5_0":
+		bytesPerElement = 0.625
+	case "iq4_nl":
+		bytesPerElement = 0.6 // 3/4 of fp16
+	case "q4_1":
+		bytesPerElement = 0.55
+	case "q4_0":
```

Comment:
Similar to below comment is there a 4-bit quantization option that's ahead of the pack in terms of compatibility, reliability and performance?

Reply:
4-bit K/V cache is the worst quality option of those available, it can have a noticeable impact on quality generations with some models. 8-bit should be considered the safe go-to.

Reply:
That makes sense, thanks. Out of all these q4 quantizations though, which do you think would be the best candidate for a default?

Reply:
@sammcj for this PR could we standardize on `q4_0` and `q8_0`?

Reply:
That's a shame but OK.

Reply:
Removed as requested.